lib/migrate.js

```diff
@@ -0,0 +1,448 @@
+ * @param {Object} newLBIndexes returned from juggler's modelDefinition.indexes()
+ * @param {Boolean} isMigrate flag to tell do we want to compare new/old indexes
+ * @callback {Function} cb The callback function
+ * @param {Object} result indexes to modify in the following format:
+ * - result: {
+ *     indexesToAdd: {
+ *       foo_index: [{foo: 'asc'}],
+ *       bar_index: [{bar1: 'asc'}, {bar2: 'asc'}] 
+ *     }, 
+ *     indexesToDrop: {
+ *       foobar_index: {
+ *         ddoc: '_design/LBModel__Foo__LBIndex__foobar_index',
+ *         fields: [{foo: 'asc'}, {bar: 'asc'}]
+ *       }
+ *     }
+ *   }
+ */
+  CouchDB.prototype.getModifyIndexes = function(mo, newLBIndexes,  isMigrate, cb) {
+    debug('CouchDB.prototype.getModifyIndexes');
+    assert(typeof newLBIndexes === 'object', 'indexes from modelDef must be an object!');
+    var self = this;
+    var results = {};
+    // `newLBIndexes` is generated from modelDef, convert it to the format we need and
+    // store in `newIndexes`
+    var newIndexes = {};
+    var modelName = mo.mo.model.modelName;
+
+    var newModelIndexes = _.pickBy(newLBIndexes, function(value, index) {
+      if (value.keys)
+        return newLBIndexes[index];
+    });
+    var newPropertyIndexes = _.pickBy(newLBIndexes, function(value, index) {
+      if (!value.keys)
+        return newLBIndexes[index];
+    });
+
+    newIndexes = _.merge(newIndexes, generateModelLevelIndexes(newModelIndexes));
+    newIndexes = _.merge(newIndexes, generatePropertyLevelIndexes(newPropertyIndexes));
+
+    // Call `getModelIndexes` to get existing indexes.
+    self.getModelIndexes(modelName, function(err, oldIndexes) {
+      if (err) return cb(err);
+      if (isMigrate) {
+        results.indexesToAdd = newIndexes;
+        results.indexesToDrop = oldIndexes;
+      } else {
+        results = self.compare(newIndexes, oldIndexes);
+      }
+      cb(null, results);
+    });
+
+    function generatePropertyLevelIndexes(indexes) {
+      var results = {};
+      for (var key in indexes) {
+        var field = {};
+        // By default the order will be `asc`,
+        // please create Model level index if you need `desc`
+        field[key.split('_index')[0]] = 'asc';
+        var fields = [field];
+        results[key] = fields;
+      };
+      return results;
+    }
+
+    function generateModelLevelIndexes(indexes) {
+      var results = {};
+      for (var key in indexes) {
+        var keys = indexes[key].keys;
+        assert(keys && typeof keys === 'object',
+          'the keys in your model index is not well defined!' +
+          'please check documentation: define index in loopback model');
+        var fields = [];
+        _.forEach(keys, function(value, key) {
+          var obj = {};
+          var order;
+          if (keys[key] === 1) order = 'asc';
+          else order = 'desc';
+          obj[key] = order;
+          fields.push(obj);
+        });
```

Comment:
See my response further down; this shouldn't be an error since it could have potentially catastrophic consequences for any migrations or updates that are in-progress. We should do our best to gracefully handle the situation and provide information on what to do to avoid that situation next time.